MusicNotationKit/MusicNotationKit/Measure.swift

```diff
@@ -53,50 +55,327 @@ public struct Measure: ImmutableMeasure, Equatable {
+        if index == 0 && notes[setIndex].count == 0 {
+            append(noteCollection, inSet: setIndex)
+            return
+        }
+
+        // If the index points to the end of the note collection, then
+        // append entry.
+        guard index != noteCount[setIndex] else {
+            append(noteCollection, inSet: setIndex)
+            return
+        }
+
+        var newMeasure = self
+        let collectionIndex = try newMeasure.noteCollectionIndex(fromNoteIndex: index, inSet: setIndex)
+
+        guard collectionIndex.tupletIndex == nil else {
+            throw MeasureError.invalidTupletIndex
+        }
+
+        if !skipTieConfig {
+            try newMeasure.prepTiesForInsertion(at: index, inSet: setIndex)
+        }
+        newMeasure.notes[setIndex].insert(noteCollection, at: collectionIndex.noteIndex)
+        self = newMeasure
+    }
+
+    internal mutating func insert(_ noteCollections: [NoteCollection], at index: Int, inSet setIndex: Int = 0, skipTies skipTieConfig: Bool) throws {
+        var newMeasure = self
+        for noteCollection in noteCollections.reversed() {
+            try newMeasure.insert(noteCollection, at: index, inSet: setIndex, skipTies: skipTieConfig)
+        }
+        self = newMeasure
+    }
+
+    internal mutating func replaceNote(at collectionIndex: NoteCollectionIndex, with noteCollection: NoteCollection, inSet setIndex: Int) throws {
         guard let tupletIndex = collectionIndex.tupletIndex else {
-            notes[setIndex][collectionIndex.noteIndex] = note
+            notes[setIndex][collectionIndex.noteIndex] = noteCollection
             return
         }
 
-        guard var tuplet = notes[setIndex][collectionIndex.noteIndex] as? Tuplet  else {
+        guard var tuplet = notes[setIndex][collectionIndex.noteIndex] as? Tuplet else {
             assertionFailure("note collection should be tuplet, but cast failed")
             throw MeasureError.internalError
         }
-        try tuplet.replaceNote(at: tuplet.flatIndexes[tupletIndex], with: note)
+        try tuplet.replaceNote(at: tuplet.flatIndexes[tupletIndex], with: noteCollection)
         notes[setIndex][collectionIndex.noteIndex] = tuplet
     }
 
-    public mutating func addNote(_ note: Note, inSet setIndex: Int = 0) {
-        notes[setIndex].append(note)
-        noteCount[setIndex] += note.noteCount
-    }
+    internal mutating func replaceNote(at collectionIndex: NoteCollectionIndex, with noteCollections: [NoteCollection], inSet setIndex: Int) throws {
+        guard collectionIndex.tupletIndex == nil || collectionIndex.tupletIndex == 0 else {
+            throw MeasureError.invalidTupletIndex
+        }
 
-    public mutating func insertNote(_ note: Note, at index: Int, inSet setIndex: Int = 0) throws {
-        // TODO: Implement
+        for note in noteCollections.reversed() {
+            try replaceNote(at: collectionIndex, with: note, inSet: setIndex)
+        }
     }
```

Comment:
This seems totally wrong upon inspection by @migue48 and I. It seems like maybe it must not be used since a test should've caught it. Please look into that.